OTAnalytics/plugin_intersect/shapely/create_intersection_events.py

```diff
@@ -0,0 +1,307 @@
+            [
+                self._apply_offset(
+                    detection.x, detection.y, detection.w, detection.h, offset
+                )
+                for detection in track.detections
+            ]
+        )
+
+    def create_line_segments(self, geometry: LineString) -> Iterable[LineString]:
+        line_segments: list[LineString] = []
+
+        for _current, _next in zip(geometry.coords[0:-1], geometry.coords[1:]):
+            line_segments.append(LineString([_current, _next]))
+        return line_segments
+
+
+class ShapelyTrackLookupTable:
+    def __init__(
+        self,
+        lookup_table: dict[TrackId, LineString],
+        geometry_builder: GeometryBuilder[LineString, Polygon],
+        offset: RelativeOffsetCoordinate,
+    ):
+        self._table = lookup_table
+        self._geometry_builder = geometry_builder
+        self._offset = offset
+
+    def look_up(self, track: Track) -> LineString:
+        if line := self._table.get(track.id):
+            return line
+        new_line = self._geometry_builder.create_track(track, self._offset)
+
+        self._table[track.id] = new_line
+        return new_line
+
+
+class ShapelyIntersectBySmallestTrackSegments(Intersector):
+    """
+    Implements the intersection strategy by splitting up the track in its smallest
+    segments and intersecting each of them with the section.
+
+    The smallest segment of a track is to generate a Line with the coordinates of
+    two neighboring detections in the track.
+
+    """
+
+    def __init__(
+        self,
+        geometry_builder: GeometryBuilder[LineString, Polygon],
+        track_lookup_table: ShapelyTrackLookupTable,
+        calculate_direction_vector_: Callable[
+            [float, float, float, float], DirectionVector2D
+        ] = calculate_direction_vector,
+    ) -> None:
+        self._geometry_builder = geometry_builder
+        self._calculate_direction_vector = calculate_direction_vector_
+        self._track_table: ShapelyTrackLookupTable = track_lookup_table
+        self._segment_lookup_table: dict[TrackId, Iterable[LineString]] = {}
+
+    def _lookup_segment(self, track: Track) -> Iterable[LineString]:
+        if segments := self._segment_lookup_table.get(track.id):
+            return segments
+        track_geometry = self._track_table.look_up(track)
+        new_segments = self._geometry_builder.create_line_segments(track_geometry)
+        self._segment_lookup_table[track.id] = new_segments
+        return new_segments
+
+    def intersect(
+        self,
+        tracks: Iterable[Track],
+        section: Section,
+        event_builder: EventBuilder,
+    ) -> list[Event]:
+        events: list[Event] = []
+        section_geometry: LineString = self._geometry_builder.create_section(section)
+        for track in tracks:
+            track_geometry = self._track_table.look_up(track)
+            if not track_geometry.intersects(section_geometry):
+                continue
```

Comment:
After merging OP#3493. We can reuse the intersection_repository to speed up event creation.

Reply:
Sounds good! I'll implement the re-use of the `intersection_repository` once  #385 has been merged~